Utils/azuretools-core/src/com/microsoft/azuretools/ui/embeddedbrowser/JxBrowserUtil.java

```diff
@@ -22,42 +22,303 @@
+     */
+    private static JComponent createBrowserViewAndLoadURL(@NotNull String url, String targetPath) throws JxBrowserException {
         Object browserView = null;
         try {
-            Class<?> browserClass = Class.forName("com.teamdev.jxbrowser.chromium.Browser");
+            Path jxBrowserJarPath = Paths.get(targetPath, getJxBrowserJarFileName());
+            Path licenseJarPath = Paths.get(targetPath, JXBROWSER_LICENSE_FILE);
+            Path commonJarPath = Paths.get(targetPath, JXBROWSER_COMMON_JAR);
+            File jxBrowserJarFile = jxBrowserJarPath.toFile();
+            File licenseJarFile = licenseJarPath.toFile();
+            File commonJarFile = commonJarPath.toFile();
+            if (!licenseJarFile.exists() || !licenseJarFile.isFile() || !jxBrowserJarFile.exists() || !jxBrowserJarFile.isFile() || !commonJarFile.exists() || !commonJarFile.isFile()) {
+                FileUtils.deleteQuietly(jxBrowserJarFile);
+                FileUtils.deleteQuietly(licenseJarFile);
+                FileUtils.deleteQuietly(commonJarFile);
+                downloadFilesFromAzure(new ArrayList<Path>(){{
+                    add(jxBrowserJarPath);
+                    add(commonJarPath);
+                    add(licenseJarPath);
+                }});
+            }
+
+            URL[] urls = new URL[] {commonJarFile.toURI().toURL(), jxBrowserJarFile.toURI().toURL(), licenseJarFile.toURI().toURL()};
+
+            ClassLoader classLoader = new URLClassLoader(urls, JxBrowserUtil.class.getClassLoader());
+
+            Class<?> browserClass = Class.forName(JXBROWSER_CLASS_BROWSER, true, classLoader);
             Object browser = browserClass.newInstance();
-            Class<?> browserViewClass = Class.forName("com.teamdev.jxbrowser.chromium.swing.BrowserView");
+            Class<?> browserViewClass = Class.forName(JXBROWSER_CLASS_BROWSERVIEW, true, classLoader);
             Constructor<?> browserViewConstructor = browserViewClass.getConstructor(browserClass);
             browserView = browserViewConstructor.newInstance(browser);
 
             if (browserView != null) {
                 Method loadURLMethod = browserClass.getMethod("loadURL", String.class);
-                loadURLMethod.invoke(url);
+                loadURLMethod.invoke(browser, url);
             }
         } catch (Exception e) {
-            throw new Exception("Fail to load JxBrowser or load URL: " + e.getMessage());
+            throw new JxBrowserException("Fail to load JxBrowser or load URL: " + e.getMessage());
         }
 
         return (JComponent)browserView;
     }
+
+    /**
+     * Download JxBrowser jars (i.e. jxbrowser-win32-6.16.jar and license.jar) and check digest (with retries) into given targetPath
+     * @param filePathNames
+     * @return
+     */
+    private static boolean downloadFilesFromAzure(@NotNull List<Path> filePathNames) {
+        boolean result = true;
+        ExecutorService threadPool = Executors.newFixedThreadPool(2);
```

Comment:
>Executors.newFixedThreadPool(2); [](start = 37, length = 32)

 maybe we can add more threads here?

Reply:
For JxBrowser, it is one large file with several small one (40MB, and several less than 1MB). The delay is mainly in the large file so 2 might be enough

---
In reply to: [156552342](https://github.com/Microsoft/azure-tools-for-java/pull/1296#discussion_r156552342) [](ancestors = 156552342)